contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("list_validators_retry"))
+        .add_message(msg))
+}
+
+fn ack_stake(deps: DepsMut, staker: String, validator: String, amount: Uint128) -> Result<IbcBasicResponse, ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Ok(IbcBasicResponse::new().add_event(Event::new("ack_stake_zero_amount")));
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED.load(deps.storage, (&staker, &validator))?;
+    stake.stake_validator(&mut val, amount);
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
+        r.stake_reset();
+        Ok(r)
+    })?;
+
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("ack_stake")))
 }
 
 pub fn fail_stake(
-    _deps: DepsMut,
-    // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let mut retries = RETRIES.load(deps.storage)?;
+    if !retries.stake_should_retry() {
+        retries.stake_reset();
+        RETRIES.save(deps.storage, &retries)?;
+        return Ok(IbcBasicResponse::new().add_event(Event::new("fail_stake")))
+    }
+    RETRIES.save(deps.storage, &retries)?;
+
+    let packet = ProviderMsg::Stake {
+        validator,
+        amount,
+        key: staker.into(),
+    };
+    let msg = IbcMsg::SendPacket {
+        channel_id: CHANNEL.load(deps.storage)?,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("retry_stake"))
+        .add_message(msg))
+}
+
+pub fn ack_unstake(
+    deps: DepsMut,
+) -> Result<IbcBasicResponse , ContractError> {
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
+        r.unstake_reset();
+        Ok(r)
+    })?;
+    Ok(IbcBasicResponse::new().add_event(Event::new("ack_unstake")))
 }
 
 pub fn fail_unstake(
-    _deps: DepsMut,
-    // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: unrelease the bonded stake, remove claim
```

Comment:
I would make this a no-op.
No retries, no action needed.
Same as `ack_unstake`